### PR TITLE
PR: Several fixes for new Editor windows

### LINKED ISSUE #20101: KeyError when updating contents in the Outline, which prevents opening new Editor windows
### What steps will reproduce the problem?

i closed the IPython Console tab and now im not able to open it

### Traceback
```python-traceback
Traceback (most recent call last):
  File "C:\Users\dgbhu\AppData\Local\Programs\Spyder\pkgs\spyder\plugins\outlineexplorer\widgets.py", line 550, in update_editor
    if not self.starting[language]:
KeyError: 'python'
```

## Versions

* Spyder version: 5.4.0  (standalone)
* Python version: 3.8.10 64-bit
* Qt version: 5.15.2
* PyQt5 version: 5.15.7
* Operating System: Windows 10

### Dependencies

```
# Mandatory:
atomicwrites >=1.2.0                 :  1.4.1 (OK)
chardet >=2.0.0                      :  5.0.0 (OK)
cloudpickle >=0.5.0                  :  2.2.0 (OK)
cookiecutter >=1.6.0                 :  2.1.1 (OK)
diff_match_patch >=20181111          :  20200713 (OK)
intervaltree                         :  None (OK)
IPython >=7.31.1;<8.0.0              :  7.34.0 (OK)
jedi >=0.17.2;<0.19.0                :  0.18.1 (OK)
jellyfish >=0.7                      :  0.9.0 (OK)
jsonschema >=3.2.0                   :  4.17.0 (OK)
keyring >=17.0.0                     :  23.9.3 (OK)
nbconvert >=4.0                      :  7.2.3 (OK)
numpydoc >=0.6.0                     :  1.5.0 (OK)
paramiko >=2.4.0                     :  2.11.0 (OK)
parso >=0.7.0;<0.9.0                 :  0.8.3 (OK)
pexpect >=4.4.0                      :  4.8.0 (OK)
pickleshare >=0.4                    :  0.7.5 (OK)
psutil >=5.3                         :  5.9.3 (OK)
pygments >=2.0                       :  2.13.0 (OK)
pylint >=2.5.0;<3.0                  :  2.15.5 (OK)
pylint_venv >=2.1.1                  :  2.3.0 (OK)
pyls_spyder >=0.4.0                  :  0.4.0 (OK)
pylsp >=1.6.0;<1.7.0                 :  1.6.0 (OK)
pylsp_black >=1.2.0                  :  1.2.1 (OK)
qdarkstyle >=3.0.2;<3.1.0            :  3.0.3 (OK)
qstylizer >=0.2.2                    :  0.2.2 (OK)
qtawesome >=1.2.1                    :  1.2.1 (OK)
qtconsole >=5.4.0;<5.5.0             :  5.4.0 (OK)
qtpy >=2.1.0                         :  2.2.1 (OK)
rtree >=0.9.7                        :  1.0.1 (OK)
setuptools >=49.6.0                  :  65.5.1 (OK)
sphinx >=0.6.6                       :  5.3.0 (OK)
spyder_kernels >=2.4.0;<2.5.0        :  2.4.0 (OK)
textdistance >=4.2.0                 :  4.5.0 (OK)
three_merge >=0.1.1                  :  0.1.1 (OK)
watchdog                             :  2.1.9 (OK)
zmq >=22.1.0                         :  24.0.1 (OK)

# Optional:
cython >=0.21                        :  0.29.32 (OK)
matplotlib >=3.0.0                   :  3.6.2 (OK)
numpy >=1.7                          :  1.22.4 (OK)
pandas >=1.1.1                       :  1.5.1 (OK)
scipy >=0.17.0                       :  1.9.3 (OK)
sympy >=0.7.3                        :  1.11.1 (OK)

# Spyder plugins:
spyder_terminal.terminalplugin 1.2.2 :  1.2.2 (OK)
```

### LINKED ISSUE #20144: RuntimeError after closing Editor window renders Editor unusable
### What steps will reproduce the problem?

Still this problem exists in Spyder 5.4.0 and I it appeares everything I do...

### Traceback
```python-traceback
Traceback (most recent call last):
  File "C:\Program Files\Spyder\pkgs\spyder\plugins\editor\plugin.py", line 1846, in todo_results_changed
    other_editorstack.set_todo_results(filename, results)
  File "C:\Program Files\Spyder\pkgs\spyder\plugins\editor\widgets\editor.py", line 2191, in set_todo_results
    self.data[index].set_todo_results(todo_results)
  File "C:\Program Files\Spyder\pkgs\spyder\plugins\editor\widgets\editorstack_helpers.py", line 186, in set_todo_results
    self.editor.process_todo(results)
  File "C:\Program Files\Spyder\pkgs\spyder\plugins\editor\widgets\codeeditor.py", line 3363, in process_todo
    for data in self.blockuserdata_list():
  File "C:\Program Files\Spyder\pkgs\spyder\plugins\editor\widgets\codeeditor.py", line 678, in blockuserdata_list
    block = self.document().firstBlock()
RuntimeError: wrapped C/C++ object of type CodeEditor has been deleted
```

## Versions

* Spyder version: 5.4.0  (standalone)
* Python version: 3.8.10 64-bit
* Qt version: 5.15.2
* PyQt5 version: 5.15.7
* Operating System: Windows 10

### Dependencies

```
# Mandatory:
atomicwrites >=1.2.0                 :  1.4.1 (OK)
chardet >=2.0.0                      :  5.0.0 (OK)
cloudpickle >=0.5.0                  :  2.2.0 (OK)
cookiecutter >=1.6.0                 :  2.1.1 (OK)
diff_match_patch >=20181111          :  20200713 (OK)
intervaltree                         :  None (OK)
IPython >=7.31.1;<8.0.0              :  7.34.0 (OK)
jedi >=0.17.2;<0.19.0                :  0.18.1 (OK)
jellyfish >=0.7                      :  0.9.0 (OK)
jsonschema >=3.2.0                   :  4.17.0 (OK)
keyring >=17.0.0                     :  23.9.3 (OK)
nbconvert >=4.0                      :  7.2.3 (OK)
numpydoc >=0.6.0                     :  1.5.0 (OK)
paramiko >=2.4.0                     :  2.11.0 (OK)
parso >=0.7.0;<0.9.0                 :  0.8.3 (OK)
pexpect >=4.4.0                      :  4.8.0 (OK)
pickleshare >=0.4                    :  0.7.5 (OK)
psutil >=5.3                         :  5.9.3 (OK)
pygments >=2.0                       :  2.13.0 (OK)
pylint >=2.5.0;<3.0                  :  2.15.5 (OK)
pylint_venv >=2.1.1                  :  2.3.0 (OK)
pyls_spyder >=0.4.0                  :  0.4.0 (OK)
pylsp >=1.6.0;<1.7.0                 :  1.6.0 (OK)
pylsp_black >=1.2.0                  :  1.2.1 (OK)
qdarkstyle >=3.0.2;<3.1.0            :  3.0.3 (OK)
qstylizer >=0.2.2                    :  0.2.2 (OK)
qtawesome >=1.2.1                    :  1.2.1 (OK)
qtconsole >=5.4.0;<5.5.0             :  5.4.0 (OK)
qtpy >=2.1.0                         :  2.2.1 (OK)
rtree >=0.9.7                        :  1.0.1 (OK)
setuptools >=49.6.0                  :  65.5.1 (OK)
sphinx >=0.6.6                       :  5.3.0 (OK)
spyder_kernels >=2.4.0;<2.5.0        :  2.4.0 (OK)
textdistance >=4.2.0                 :  4.5.0 (OK)
three_merge >=0.1.1                  :  0.1.1 (OK)
watchdog                             :  2.1.9 (OK)
zmq >=22.1.0                         :  24.0.1 (OK)

# Optional:
cython >=0.21                        :  0.29.32 (OK)
matplotlib >=3.0.0                   :  3.6.2 (OK)
numpy >=1.7                          :  1.22.4 (OK)
pandas >=1.1.1                       :  1.5.1 (OK)
scipy >=0.17.0                       :  1.9.3 (OK)
sympy >=0.7.3                        :  1.11.1 (OK)

# Spyder plugins:
spyder_terminal.terminalplugin 1.2.2 :  1.2.2 (OK)
```

### PR DESCRIPTION
**Pull request by ccordoba12**
## Description of Changes

- Remove EditorStacks present on EditorMainWindow when it's closed. We were not removing them from the list of EditorStacks maintained by the plugin, which made the Editor unusable by generating endless RuntimeErrors. 
- Fix Outline displayed on EditorMainWindow, which was not showing any symbol.
- Update Outline for cloned editors after editing text in them (this was not working).
- Update symbols when switching between cloned editors (this was not working either).
- Don't update Outline if EditorMainWindow or MainWindow are minimized (to avoid lags on the main interface if they are not visible).
- Add `sig_mainwindow_state_changed` to plugins so they can take different actions when the main window is maximized/minimized.
- Expand tests to check Outline is working on EditorMainWindow.

### Issue(s) Resolved

Fixes #20144
Fixes #20101

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by ccordoba12** [comment]
> Checking the failing tests maybe the error is related with NumPy latest release (1.24.0) and the removal of things that were deprecated?

Good guess! That's right, Numpy removed `np.bool`, which we were using in some tests (I'm fixing that in PR #20224; after it's merged, I'll rebase this one).

> Other than the failing tests (which most probably are unrelated with the changes done here) this LGTM +1

A test that I expanded to cover these fixes is hanging, so I still have to check what's happening with it. I'll give you another ping when I manage to make it pass.

**Comment by ccordoba12** [comment]
@dalthviz, this should be ready now.

**Review by dalthviz — COMMENTED** [review]
Thanks @ccordoba12 , checking locally I got a strange behavior with this. After opening a project (Spyder source code) I splitted the Editor window and then used the `New window` option. After that the outline explorer kept showing the spinner in the Outline Explorer:

imagen


Then, >I tried closing the Editor window and restart the LSP server (to see if that helped with the spinner still running) but that triggered an error:

```
Traceback (most recent call last):
  File "e:\acer\documentos\spyder\spyder otros\carlos\spyder\spyder\plugins\editor\plugin.py", line 401, in register_completion_capabilities
    self.start_completion_services(language)
  File "e:\acer\documentos\spyder\spyder otros\carlos\spyder\spyder\plugins\editor\plugin.py", line 406, in start_completion_services
    editorstack.start_completion_services(language)
  File "e:\acer\documentos\spyder\spyder otros\carlos\spyder\spyder\plugins\editor\widgets\editor.py", line 1715, in start_completion_services
    editor.start_completion_services()
  File "e:\acer\documentos\spyder\spyder otros\carlos\spyder\spyder\plugins\editor\widgets\codeeditor.py", line 1119, in start_completion_services
    self.document_did_open()
  File "e:\acer\documentos\spyder\spyder otros\carlos\spyder\spyder\plugins\completion\decorators.py", line 20, in wrapper
    params = req(self, *args, **kwargs)
  File "e:\acer\documentos\spyder\spyder otros\carlos\spyder\spyder\plugins\editor\widgets\codeeditor.py", line 1191, in document_did_open
    self._timer_sync_symbols_and_folding.timeout.connect(
TypeError: connection is not unique
```

I was unable to reproduce again the issue with the spinner not stopping but the error when trying to restart the LSP persisted (although not sure if related with the changes done here)

**Comment by ccordoba12** [comment]
> I was unable to reproduce again the issue with the spinner not stopping

Ok, thanks for the extra testing @dalthviz, I was unable to reproduce that on my end either. So I think we should merge this one for now and wait to see if users are able to give us a series of steps that reproduce this error.

> but the error when trying to restart the LSP persisted (although not sure if related with the changes done here)

I was able to reproduce that error while testing different solutions for issue #19565 and I fixed it on PR #20264. Please check if it also solves the problem for you.

File: spyder/plugins/editor/widgets/editor.py
Comment (dalthviz) [inline spyder/plugins/editor/widgets/editor.py:3429]: Maybe adding why the `+ 20` is needed/was approximated could be worthy?

**Review by dalthviz — APPROVED** [review]
Checked locally on Windows and seems like everything is working :+1:

Left a suggestion for a comment which maybe has a typo and suggestion/comment to add a comment regarding a value when doing a validation to make visible or not the outlienexplorer tree widget.

Other than that this LGTM :+1:

**Comment by ccordoba12** [comment]
Thanks @dalthviz for your last review. I implemented your suggestions, so this ready to be merged.

**Pull request merged**
